tests/system/test_system_dbapi.py

```diff
@@ -426,6 +427,28 @@ def test_DDL_commit(self):
         cur.execute("DROP TABLE Singers")
         conn.commit()
 
+    def test_read_only(self):
+        """
+        Check that connection set to `read_only=True` uses
+        ReadOnly transactions.
+        """
+        conn = Connection(Config.INSTANCE, self._db, read_only=True)
+        cur = conn.cursor()
+
+        with self.assertRaises(ProgrammingError):
+            cur.execute(
+                """
+    UPDATE contacts
+    SET first_name = 'updated-first-name'
+    WHERE first_name = 'first-name'
+    """
+            )
+
+        cur.execute("SELECT * FROM contacts")
+
+        with self.assertRaises(exceptions.FailedPrecondition):
+            conn.commit()
```

Comment:
This should not raise an exception. This should remove the read-only transaction from the connection and return the session to the pool. If there is no possibility of 'committing' (or otherwise ending) a read-only transaction, the connection will __never__ see any new data after the read-only transaction has started.

Reply:
It is something that comes from the backend with a message:
`400 Cannot commit or rollback a read-only or partitioned-dml transaction`. So, it's not us who raise this exception.